binance-pipeline/lambdas/binance_historical_extraction.py

```diff
@@ -0,0 +1,249 @@
+# ./lambdas/binance_historical_extraction.py
+from datetime import datetime
+import json
+import math
+import time
+#
+from binance import BinanceQueryAPI
+from binance import TradeIDFinder, get_latest_trade_id, get_first_trade_id
+
+CHUNK_LIMIT = 1e3
+DATEFORMAT = "%Y-%m-%d"
+RATE_LIMIT: int = 12e3
+RESERVED_CONCURRENCY = 4
+SQS_QUEUE_NAME_EXTRACTION = ''
+SQS_QUEUE_NAME_DELAY_DAY = ''
+SQS_QUEUE_NAME_DELAY_HOUR = ''
+SQS_QUEUE_NAME_DELAY_MINUTE = ''
+SQS_QUEUE_NAME_DELAY_SECOND = ''
+
+def parse_date(date: str) -> datetime:
+
+    return datetime.strptime(date, DATEFORMAT)
+
+
+def parse_event(payload: dict):
+    if (type_ := payload['type']) == 'orchestration':
+        # user sent configuration
+        handle_orchestration(payload['config'])
+    elif type_ == 'api_call':
+        # sqs sent the payload to send as GET to API
+        handle_api_call(payload['config'])
+    else:
+        raise Exception("lambda only accepts api_call or orchestration")
+
+    return
+
+
+def ensure_sqs_trigger(record):
+    if (event_source := record["eventSource"]) != "aws:sqs":
+        raise Exception(f"Event source must be aws:sqs not {event_source = }")
+
+    return
+
+
+def handle_orchestration(config: dict):
+    start_date: str = config.get('start_date')
+    end_date: str = config.get('end_date')
+    reverse: bool = config.get('reverse')
+    symbol: str = config.get('symbol')
+
+    if reverse:
+        handle_reverse_load(symbol, start_date)
+    else:
+        if not start_date:
+            handle_from_start(symbol, end_date)
+        else:
+            handle_from_date(symbol, start_date, end_date)
```

Comment:
Can `handle_from_date` handle when `start_date` is `None` or defaulted? Then this can be collapsed into a single invocation.

Reply:
I would like a better way of diverting these handlers, a factory pattern feels right for this and is worth a refactor

Is your point that handle_from_date could take a start_date of None and thus be from trade_id=0, removing the need for a handle_from_start method at all? I think that's quite possible, the handle_from_date is probably already unnecessarily complex (and I haven't even implemented the end_date functionality yet). 

The whole pattern needs to be improved, I will come up with a chart and then see how best to do this

Reply:
My thinking is that handle_from_start is a special case of handle_from_date, so more to collapse them into a single pattern/idiom.